src/Data/Parameterized/FinMap/Unsafe.hs

```diff
@@ -0,0 +1,218 @@
+--
+-- * Its keys must be less than the nat in its type.
+-- * Its size must be less than the maximum Int.
+--
+-- If these invariants hold, all of the unsafe operations in this module
+-- (fromJust, unsafeCoerce) will work as intended.
+
+-- | @'FinMap' n a@ is a map with @'Fin' n@ keys and @a@ values.
+newtype FinMap (n :: Nat) a = FinMap { getFinMap :: IntMap a }
+
+instance Eq a => Eq (FinMap n a) where
+  fm1 == fm2 = getFinMap fm1 == getFinMap fm2
+  {-# INLINABLE (==) #-}
+
+instance Functor (FinMap n) where
+  fmap f = FinMap . fmap f . getFinMap
+  {-# INLINABLE fmap #-}
+
+instance Foldable (FinMap n) where
+  foldMap f = foldMap f . getFinMap
+  {-# INLINABLE foldMap #-}
+
+instance FunctorWithIndex (Fin n) (FinMap n) where
+  imap f = FinMap . IntMap.mapWithKey (f . unsafeFin) . getFinMap
+  -- Inline so that RULES for IntMap.mapWithKey can fire
+  {-# INLINE imap #-}
+
+instance FoldableWithIndex (Fin n) (FinMap n) where
+  ifoldMap f = IntMap.foldMapWithKey (f . unsafeFin) . getFinMap
+
+-- | Non-lawful instance, provided for testing
+instance Show a => Show (FinMap n a) where
+  show fm = show (getFinMap fm)
+  {-# INLINABLE show #-}
+
+------------------------------------------------------------------------
+-- Query
+
+-- | /O(1)/. Is the map empty?
+null :: FinMap n a -> Bool
+null = IntMap.null . getFinMap
+{-# INLINABLE null #-}
+
+-- | /O(min(n,W))/. Fetch the value at the given key in the map.
+lookup :: Fin n -> FinMap n a -> Maybe a
+lookup k = IntMap.lookup (unsafeFinToInt k) . getFinMap
+{-# INLINABLE lookup #-}
+
+-- | Unsafely create a @'Fin' n@ from an 'Int' which is known to be less than
+-- @n@ for reasons not visible to the type system.
+unsafeFin :: forall n. Int -> Fin n
+unsafeFin i =
+  case NatRepr.mkNatRepr (intToNat i) of
+    Some (repr :: NatRepr m) ->
+      case unsafeCoerce (NatRepr.LeqProof :: LeqProof 0 0) :: LeqProof (m + 1) n of
+        NatRepr.LeqProof -> mkFin @m @n repr
+
+-- | /O(n)/. Number of elements in the map.
+size :: forall n a. FinMap n a -> Fin (n + 1)
+size = unsafeFin . IntMap.size . getFinMap
+{-# INLINEABLE size #-}
+
+------------------------------------------------------------------------
+-- Construction
+
+-- | /O(1)/. Increase maximum key/size by 1.
+--
+-- Requires @n + 1 < (maxBound :: Int)@.
+incMax :: FinMap n a -> FinMap (n + 1) a
+incMax = FinMap . getFinMap
+{-# INLINE incMax #-}
+
+-- | /O(1)/. Increase maximum key/size.
+--
+-- Requires @m < (maxBound :: Int)@.
+embed :: (n <= m) => NatRepr m -> FinMap n a -> FinMap m a
+embed _ = FinMap . getFinMap
+{-# INLINE embed #-}
+
+-- | /O(1)/. The empty map.
+empty :: KnownNat n => FinMap n a
+empty = FinMap IntMap.empty
+{-# INLINE empty #-}
+
+-- | /O(1)/. A map with one element.
+singleton :: a -> FinMap 1 a
+singleton = FinMap . IntMap.singleton 0
+{-# INLINABLE singleton #-}
+
+-- | /O(min(n,W))/.
+insert :: Fin n -> a -> FinMap n a -> FinMap n a
+insert k v = FinMap . IntMap.insert (unsafeFinToInt k) v . getFinMap
```

Comment:
`insert k = fmap (IntMap.insert (unsafeFinToInt k))`  ?

Reply:
`FinMap` is a functor, but its value type is `a` rather than `IntMap a`, so this formulation doesn't quite work:
```
src/Data/Parameterized/FinMap/Unsafe.hs:168:12: error:
    • Couldn't match type ‘IntMap a0 -> IntMap a0’ with ‘FinMap n a’
      Expected type: a -> FinMap n a -> FinMap n a
        Actual type: (FinMap n a -> a0)
                     -> FinMap n a -> IntMap a0 -> IntMap a0
    • Probable cause: ‘fmap’ is applied to too few arguments
      In the expression: fmap (IntMap.insert (unsafeFinToInt k))
      In an equation for ‘insert’:
          insert k = fmap (IntMap.insert (unsafeFinToInt k))
    • Relevant bindings include
        k :: Fin n (bound at src/Data/Parameterized/FinMap/Unsafe.hs:168:8)
        insert :: Fin n -> a -> FinMap n a -> FinMap n a
          (bound at src/Data/Parameterized/FinMap/Unsafe.hs:168:1)
    |
168 | insert k = fmap (IntMap.insert (unsafeFinToInt k))
    | 
```

Reply:
Ahh, right.